### ISSUE TITLE
False positive with assert-if-true/if-false

### ISSUE THREAD

**Issue by klimick** [vimeo/psalm#5248]
If assert function has `@psalm-assert-if-false` and `@psalm-assert-if-true` at the same time then it works incorrectly:
https://psalm.dev/r/91745556ee

Quick workaround is drop `@psalm-assert-if-false` or `@psalm-assert-if-true` from assert function:
https://psalm.dev/r/a5500b4ed2

But above workaround doesn't work in more difficult cases:
https://psalm.dev/r/408248f46f

**Comment by weirdan** [comment]
Also this doesn't seem right: https://psalm.dev/r/65cbc43cff

**Comment by weirdan** [comment]
@muglug it looks like Psalm generates inverted assertion (`!string`) for false return here, but should it really do that?

**Comment by whsv26** [comment]
Encountered the same problem for Either monad. I'd like to do something like this, but unfortunately it's not working for both psalm-assert-if-true and psalm-assert-if-false at the same time.

```php
/**
 * @psalm-assert-if-true Left<L, R> $this
 * @psalm-assert-if-false Right<L, R> $this
 */
public function isLeft(): bool {}

/**
 * @psalm-assert-if-true Right<L, R> $this
 * @psalm-assert-if-false Left<L, R> $this
 */
public function isRight(): bool {}
```

**Comment by klimick** [comment]
Issue is solved: https://psalm.dev/r/b0b6a03bd6
It seems in the #9209

**Comment by klimick** [comment]
I hastened to close the issue. Sorry. Issue still exists.
https://psalm.dev/r/6125f6ef36
